cmd/controller/server.go

```diff
@@ -0,0 +1,73 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/golang/glog"
+	"github.com/gorilla/mux"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+const (
+	prometheusMetricsServerPort            = 9402
+	prometheusMetricsServerShutdownTimeout = 10 * time.Second
+	prometheusMetricsServerReadTimeout     = 8 * time.Second
+	prometheusMetricsServerWriteTimeout    = 8 * time.Second
+	prometheusMetricsServerMaxHeaderBytes  = 1 << 20
+)
+
+type prometheusMetricsServer struct {
+	http.Server
+}
+
+func prometheusHandler() http.Handler {
+	return prometheus.Handler()
+}
+
+func newPrometheusMetricsServer() *prometheusMetricsServer {
+
+	r := mux.NewRouter()
+	r.Handle("/metrics", prometheusHandler())
+
+	// Create server and register prometheus metrics handler
+	s := &prometheusMetricsServer{
+		Server: http.Server{
+			Addr:           fmt.Sprintf(":%v", prometheusMetricsServerPort),
+			ReadTimeout:    prometheusMetricsServerReadTimeout,
+			WriteTimeout:   prometheusMetricsServerWriteTimeout,
+			MaxHeaderBytes: prometheusMetricsServerMaxHeaderBytes,
+			Handler:        r,
+		},
+	}
+
+	return s
+}
+
+func (s *prometheusMetricsServer) WaitShutdown(stopCh <-chan struct{}) {
+	<-stopCh
+	glog.Info("Stopping Prometheus metrics server...")
+
+	ctx, cancel := context.WithTimeout(context.Background(), prometheusMetricsServerShutdownTimeout)
+	defer cancel()
+
+	if err := s.Shutdown(ctx); err != nil {
+		glog.Infof("Prometheus metrics server shutdown error: %v", err)
+		return
+	}
+
+	glog.Info("Prometheus metrics server gracefully stopped")
+}
+
+func StartPrometheusMetricsServer(stopCh <-chan struct{}) {
+	s := newPrometheusMetricsServer()
+
+	go func() {
+		glog.Infof("Listening on http://0.0.0.0%s", s.Addr)
+		glog.Fatal(s.ListenAndServe())
```

Comment:
I worry about this call to `Fatal` here. Ideally, the ungraceful termination should be propagated back to `start.go` so that the controller can also be gracefully shut down.

Also, can you rewrite to be more like:

```
if err := s.ListenAndServer(); err != nil {
    glog.Fatalf("Error running prometheus metrics server: %s", err.Error())
}

glog.Fatalf("Prometheus server unexpectedly exited")
```

Reply:
We can leave the termination propagation for now btw. Create an issue for it so we don't forget :smile:

Reply:
So when `s.Shutdown` is called, I am hoping it causes `s.ListenAndServe()` to return nil. This will then leave the program executing until the main worker loop shuts down.

We need the graceful termination path to *not* call Fatal/cause the program to exit at the end of execution, else the main control loop may be midway through cleaning up in-flight requests when Fatal is called.